### ISSUE TITLE
HX Triage timeout with graph generation

### ISSUE THREAD

**Issue by kennyleehk** [yampelo/beagle#24]
I am running Dockers on Mac with the latest pull of Beagle today.

I had a HX Triage MANS file that has created 870223 nodes.  The graph creation times out after 50 minutes.  I ran Beagle with a specification of 12 CPU cores but it did not help.

```
root# docker run --cpus=12.000 -v "/data/beagle":"/data/beagle" -p 8000:8000 -e "BEAGLE__GENERAL__LOG_LEVEL=DEBUG" yampelo/beagle
[2019-04-17 04:59:40 +0000] [1] [INFO] Starting gunicorn 19.9.0
[2019-04-17 04:59:40 +0000] [1] [INFO] Listening at: http://0.0.0.0:8000 (1)
[2019-04-17 04:59:40 +0000] [1] [INFO] Using worker: sync
[2019-04-17 04:59:40 +0000] [9] [INFO] Booting worker with pid: 9
[2019-04-17 04:59:40 +0000] [10] [INFO] Booting worker with pid: 10
[2019-04-17 04:59:40 +0000] [11] [INFO] Booting worker with pid: 11
[2019-04-17 04:59:40 +0000] [15] [INFO] Booting worker with pid: 15
[2019-04-17 04:59:40 +0000] [16] [INFO] Booting worker with pid: 16
[2019-04-17 04:59:40 +0000] [18] [INFO] Booting worker with pid: 18
[2019-04-17 04:59:41 +0000] [54] [INFO] Booting worker with pid: 54
[2019-04-17 04:59:41 +0000] [55] [INFO] Booting worker with pid: 55
[2019-04-17 04:59:41 +0000] [56] [INFO] Booting worker with pid: 56
[2019-04-17 04:59:41 +0000] [68] [INFO] Booting worker with pid: 68
[2019-04-17 04:59:41 +0000] [71] [INFO] Booting worker with pid: 71
[2019-04-17 04:59:41 +0000] [84] [INFO] Booting worker with pid: 84
2019-04-17T05:00:03 | beagle.web.api.views.new:208 | INFO | Recieved upload request for datasource=<HXTriage>, transformer=<FireEyeHXTransformer>
2019-04-17T05:00:03 | beagle.web.api.views.new:245 | INFO | Transforming data to a graph.
2019-04-17T05:00:03 | beagle.web.api.views.new:274 | INFO | Saved uploaded files {'triage': <tempfile._TemporaryFileWrapper object at 0x7f228edf3e48>}
2019-04-17T05:00:03 | beagle.datasources.hx_triage.__init__:53 | INFO | Setting up HXTriage for /tmp/tmp0xd99xy_
2019-04-17T05:00:03 | beagle.datasources.hx_triage.__init__:57 | DEBUG | Generated temporary directory /tmp/tmpr0jkjmx3_beagle
2019-04-17T05:00:04 | beagle.datasources.hx_triage.__init__:93 | DEBUG | Mapped 5pq8DVSJOP993LATmGmm3c to stateagentinspector
2019-04-17T05:00:04 | beagle.transformers.fireeye_hx_transformer.__init__:17 | INFO | Created FireEyeHX Transformer.
2019-04-17T05:00:04 | beagle.transformers.base_transformer.run:77 | DEBUG | Launching transformer
2019-04-17T05:00:04 | beagle.transformers.base_transformer.run:86 | DEBUG | Started producer thread
2019-04-17T05:00:04 | beagle.transformers.base_transformer.run:98 | DEBUG | Started 11 consumer threads
2019-04-17T05:01:10 | beagle.transformers.base_transformer._producer_thread:125 | DEBUG | Producer Thread Thread-1 finished after 301402 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-4 finished after processing 27957 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-2 finished after processing 26870 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-8 finished after processing 27376 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-7 finished after processing 29033 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-12 finished after processing 26999 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-5 finished after processing 26087 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-10 finished after processing 28742 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-3 finished after processing 27132 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-9 finished after processing 28139 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-11 finished after processing 27934 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-6 finished after processing 25144 events
2019-04-17T05:01:10 | beagle.transformers.base_transformer.run:111 | INFO | Finished processing of events, created 870223 nodes.
2019-04-17T05:01:10 | beagle.backends.networkx.__init__:51 | INFO | Initialized NetworkX Backend
2019-04-17T05:01:10 | beagle.backends.networkx.graph:65 | INFO | Beginning graph generation.
[2019-04-17 05:50:03 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:11)
[2019-04-17 05:50:03 +0000] [11] [INFO] Worker exiting (pid: 11)
[2019-04-17 05:50:04 +0000] [177] [INFO] Booting worker with pid: 177
```

I used this CURL command to submit the MANS file.  The CURL timed out after 50 minutes.

```
# curl -F 'triage=@TJmxAb8UzW8ddKQHqTKPwE.MANS' -F 'datasource=HXTriage' -F 'transformer=FireEyeHXTransformer' -F 'comment=SP Triage' http://localhost:8000/api/new

curl: (52) Empty reply from server
```

I ran another MANS file that created only 308316 nodes.  That graph was successfully generated.

```
2019-04-17T06:31:24 | beagle.web.api.views.new:208 | INFO | Recieved upload request for datasource=<HXTriage>, transformer=<FireEyeHXTransformer>
2019-04-17T06:31:24 | beagle.web.api.views.new:245 | INFO | Transforming data to a graph.
2019-04-17T06:31:24 | beagle.web.api.views.new:274 | INFO | Saved uploaded files {'triage': <tempfile._TemporaryFileWrapper object at 0x7f228edb1390>}
2019-04-17T06:31:24 | beagle.datasources.hx_triage.__init__:53 | INFO | Setting up HXTriage for /tmp/tmpy2mka_nf
2019-04-17T06:31:24 | beagle.datasources.hx_triage.__init__:57 | DEBUG | Generated temporary directory /tmp/tmpi3x4vzrw_beagle
2019-04-17T06:31:24 | beagle.datasources.hx_triage.__init__:93 | DEBUG | Mapped 95ndX1K8xy403iqPyLqAd3 to stateagentinspector
2019-04-17T06:31:24 | beagle.transformers.fireeye_hx_transformer.__init__:17 | INFO | Created FireEyeHX Transformer.
2019-04-17T06:31:24 | beagle.transformers.base_transformer.run:77 | DEBUG | Launching transformer
2019-04-17T06:31:24 | beagle.transformers.base_transformer.run:86 | DEBUG | Started producer thread
2019-04-17T06:31:24 | beagle.transformers.base_transformer.run:98 | DEBUG | Started 11 consumer threads
2019-04-17T06:31:47 | beagle.transformers.base_transformer._producer_thread:125 | DEBUG | Producer Thread Thread-1 finished after 103165 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-7 finished after processing 9035 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-3 finished after processing 10368 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-10 finished after processing 8101 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-6 finished after processing 8517 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-8 finished after processing 9343 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-12 finished after processing 8620 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-2 finished after processing 9708 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-4 finished after processing 9562 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-11 finished after processing 11418 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-9 finished after processing 9422 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer._consumer_thread:136 | DEBUG | Consumer Thread Thread-5 finished after processing 9082 events
2019-04-17T06:31:47 | beagle.transformers.base_transformer.run:111 | INFO | Finished processing of events, created 308316 nodes.
2019-04-17T06:31:47 | beagle.backends.networkx.__init__:51 | INFO | Initialized NetworkX Backend
2019-04-17T06:31:47 | beagle.backends.networkx.graph:65 | INFO | Beginning graph generation.
2019-04-17T06:31:55 | beagle.backends.networkx.graph:71 | INFO | Completed graph generation.
2019-04-17T06:31:55 | beagle.backends.networkx.graph:72 | INFO | Graph contains 6332 nodes and 39520 edges.
2019-04-17T06:31:55 | beagle.web.api.views.new:305 | INFO | Finished generating graph
2019-04-17T06:31:56 | beagle.web.api.views.new:340 | INFO | Added graph to database with id=1
2019-04-17T06:31:58 | beagle.web.api.views.new:344 | INFO | Saved graph to /data/beagle/fireeye_hx/8203c3e57725e9671bb016fa1f47d1759e92dac0f915113880327009ac3dab62.json
```

Thanks for the great work so far with  Beagle!

Regards

**Comment by yampelo** [comment]
Would you be willing to share the problematic triage? it may have some edge case causing it to hang that i haven’t encountered

**Comment by yampelo** [comment]
This should now be fixed by #62, large graphs made through the web interface (or with the `consolidate_edge` flag) finish in a matter of minutes now.
